[POS] Custom payment UI — Part 5 | Unit tests clean up

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
#### https://github.com/woocommerce/woocommerce-android/pull/12943

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
The intent of This PR is to clean up unit tests of `CardReaderPaymentControllerTest` and `CardReaderPaymentViewModelTest`.

#### This is **part 5** of 5 PRs refactoring the Payment flow:
1. [[POS] Custom payment UI — Part 1 | Separating payment collection management from Android Framework](https://github.com/woocommerce/woocommerce-android/pull/12853)
2. [[POS] Custom payment UI — Part 2 | Separating CardReaderPaymentController's events from MultiLiveEvent](https://github.com/woocommerce/woocommerce-android/pull/12870)
3. [[POS] Custom Payment UI – Part 3 | Emitting UI–agnostic payment states in CardReaderPaymentController](https://github.com/woocommerce/woocommerce-android/pull/12877)
4. [[POS] Custom payment UI — Part 4 | Switch from viewStateData to paymentState](https://github.com/woocommerce/woocommerce-android/pull/12898)
5. **[[POS] Custom payment UI — Part 5 | Unit tests clean up](https://github.com/woocommerce/woocommerce-android/pull/12943)**

💡 This PR intends to merge changes to the `feature/custom-payment-ui` feature branch, that will be additionally tested and merged to `trunk` later.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

This PR is quite big, however, it modifies only two files: `CardReaderPaymentControllerTest` and `CardReaderPaymentViewModelTest`. It also does not introduce any new tests, only moves test methods around so that they are closer to the class they are testing.

1. `ViewState` tests
I copied all the unit test methods verifying `CardReaderPaymentViewModel.viewStateData` to `CardReaderPaymentControllerTest` and adapted them to test `CardReaderPaymentController.paymentState` instead. I left original tests as well because they are testing the `CardReaderPaymentStateToViewStateMapper` as a black box.
2. Tests not testing `ViewState` were moved to `CardReaderPaymentControllerTest` (and removed from `CardReaderPaymentViewModelTest`)

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
I verified that `CardReaderPaymentViewModelTest` does not contain test methods that are more suitable for the controller.

### Testing
I recommend checking this PR as a whole
1. Review the `CardReaderPaymentViewModelTest` and look if there is a class that could be moved or copied/adapted to the `CardReaderPaymentController` class.
2. Verify that unit tests and other CI checks are passing

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
—

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when RETRY message received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when INSERT_CARD received, then collect payment hint updated`

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when SWIPE_CARD received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when REMOVE_CARD received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when TRY_OTHER_CARD message received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when CARD_REMOVED_TOO_EARLY message received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when TRY_OTHER_READ message received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given collect payment shown, when MULTIPLE_CARDS_DETECTED received, then collect payment hint updated`

samiuelson committed:
Add PaymentController test:

`given fetching order fails, when payment screen shown, then ExternalReaderFailedPayment state is shown`

samiuelson committed:
Add PaymentController test:

`given fetching order fails and tpp, when payment screen shown, then BuiltInReaderFailedPayment state is shown`

samiuelson committed:
Add PaymentController test:

`when fetching order fails, then event tracked`

samiuelson committed:
Move test from VM to PaymentController :

`given fetching order succeeds, when payment screen shown, then order currency stored `

samiuelson committed:
Add test PaymentController:

`when payment screen shown, then loading data state is emitted`

samiuelson committed:
Add test PaymentController:

`when payment not collectable, then error event emitted and flow terminated`

samiuelson committed:
Move test from VM PaymentController:

`when flow started, then correct payment description is propagated to CardReaderManager`

samiuelson committed:
Move test from VM PaymentController:

`when flow started, then correct statement descriptor is propagated to CardReaderManager`

samiuelson committed:
Add test to PaymentController:

`when initializing payment, then Loading state emitted`

samiuelson committed:
Add test to PaymentController:

`when collecting payment, then CollectingPayment state emitted`

samiuelson committed:
Add test to PaymentController:

`given built in reader,when collecting payment, then ui CollectingPayment state emitted`

samiuelson committed:
Add test to PaymentController:

`when processing payment, then ProcessingPayment state emitted`

samiuelson committed:
Add test to PaymentController:

`given built in reader, when processing payment, then ProcessingPayment state emitted`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with card present, then tracking keeper stores payment type`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with interac present, then tracking keeper stores payment type`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with interac present, then track interac success`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with card present, then do not track interac success`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with unknown type, then do not track interac success`

samiuelson committed:
Move test from VM to PaymentController:

`when processing payment completed with unknown, then tracking keeper stores payment type`

samiuelson committed:
Add PaymentController test:

`when capturing payment, then capturing payment state emitted`

samiuelson committed:
Add PaymentController test:

`given built in reader, when capturing payment, then ui updated to capturing payment state`

samiuelson committed:
Add PaymentController test:

`given billing email empty, when external payment completed, then payment successful state emitted`

samiuelson committed:
Add PaymentController test:

`given billing not empty, when external payment completed, then payment successful receipt sent state emitted`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given billing not empty, when built in payment completed, then built in payment successful receipt sent state emitted`

samiuelson committed:
Add `CardReaderPaymentController` test:

`when payment completed, then success sound is played`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`when payment completed, then event tracked`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given external reader, when payment fails, then failed state emitted`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given built in reader fails with Unknown error, when view model starts, then CTA with contact support button is provided`

`given built in reader fails with Unknown error, when view model starts, then CTA with contact support button is provided`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given external reader fails with generic error, when contact support clicked, then contact support emitted and flow canceled`

samiuelson committed:
Add `CardReaderPaymentController` test:

 `given built in reader fails with generic error, when contact support clicked, then contact support emitted and flow canceled`

samiuelson committed:
Add `CardReaderPaymentController` test:

`when payment fails, then event tracked`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given external reader, when payment fails because of AMOUNT_TOO_SMALL, then failed state is not retryable`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given external reader, when payment fails not because of AMOUNT_TOO_SMALL, then failed state is retryable`

samiuelson committed:
Add `CardReaderPaymentController` test:

 `given built in reader, when payment fails not because of AMOUNT_TOO_SMALL, then failed state has Try again button`

`given built in reader, when payment fails PIN_REQUIRED, then failed state has purchase card reader cta`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given built in reader, when purchase button clicked, then purchase even emmited`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given built in reader, when payment fails because of AMOUNT_TOO_SMALL, then clicking on ok button triggers exit event`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given payment fails because of NFC_DISABLED, when primary button clicked, then tracked and enablenfc emitted`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given user clicks on retry and built in, when payment fails and retryData are null, then flow restarted from scratch`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given failed payment and external reader, when user retries, then retryCollectPayment invoked`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given failed payment and external reader, when user retries, then flow retried with provided PaymentData`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given failed payment and built-in reader, when user retries, then flow retried with provided PaymentData`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given external failed payment, when user clicks on secondary button, then exit event is triggered`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given built in failed payment, when user clicks on secondary button, then exit event is triggered`

samiuelson committed:
Add `CardReaderPaymentController` test:

`when loading data, then cancellation is possible`

samiuelson committed:
Add `CardReaderPaymentController` test:

`when collecting payment, then cancellation is possible`

samiuelson committed:
Rename vars

samiuelson committed:
Add `CardReaderPaymentController` test:

`when payment fails, then cancellation is possible`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`when payment succeeds, then receiptUrl stored into a persistant storage`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

 `given payment flow already started, when start() is invoked, then flow is not restarted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and external, when user clicks on print receipt button, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and built in, when user clicks on print receipt button, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and external, when user clicks on print receipt button, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and built in, when user clicks on print receipt button, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and external, when user clicks on print receipt button, then printing receipt state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and built-in, when user clicks on print receipt button, then printing receipt state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and external, when user clicks on print receipt button, then printing receipt state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and built-in, when user clicks on print receipt button, then printing receipt state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and built in, when print result received, then payment successful state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and external, when print result received, then payment success receipt sent state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and built in, when print result received, then payment success receipt sent state shown`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given in printing receipt state and external, when view recreated, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given in printing receipt state and built in, when view recreated, then PrintReceipt event emitted`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given not in printing receipt state, when view recreated, then state not changed`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and external, when user clicks on print receipt button, then event tracked`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email empty and built in, when user clicks on print receipt button, then event tracked`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and external, when user clicks on print receipt button, then event tracked`

samiuelson committed:
Move test from VM to `CardReaderPaymentController`:

`given billing email not empty and built in, when user clicks on print receipt button, then event tracked`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given get receipt url fails, when user clicks on print receipt button, then error event emitted`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given get receipt url succeeds, when user clicks on print receipt button, then PrintReceipt emitted`

samiuelson committed:
Move tests to `CardReaderPaymentController`:

`when OS accepts the print request, then print success event tracked`

`when OS refuses the print request, then print failed event tracked`

`when manually cancels the print request, then print cancelled event tracked`

samiuelson committed:
Merge branch `trunk` into `custom-payment-ui-5`

samiuelson committed:
Merge branch 'custom-payment-ui-4' into custom-payment-ui-5

samiuelson committed:
Clean up code after merge

samiuelson committed:
Move tests from `CardReaderPaymentViewModelTest` to `CardReaderPaymentController`

samiuelson committed:
Move tests from `CardReaderPaymentViewModelTest` to `CardReaderPaymentController`

samiuelson committed:
Move tests from `CardReaderPaymentViewModelTest` to `CardReaderPaymentController`

samiuelson committed:
Move tests from `CardReaderPaymentViewModelTest` to `CardReaderPaymentController`

samiuelson committed:
Add `CardReaderPaymentController` test:

`given collect payment NOT shown, when show additional info event received, then event ignored`

samiuelson committed:
Move Interac flow tests to `CardReaderPaymentController`

samiuelson committed:
Move tests to `CardReaderPaymentController`

samiuelson committed:
Move tests to `CardReaderPaymentController`

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch 'feature/custom-payment-ui' into custom-payment-ui-5

samiuelson committed:
Improve formatting
